// ngx-bootstrap/typeahead/esm2015/models/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * A context for the `optionsListTemplate`
 * input template in case you want to override default one
 * @record
 */
export function TypeaheadOptionListContext() { }
if (false) {
    /**
     * All matches
     * @type {?}
     */
    TypeaheadOptionListContext.prototype.matches;
    /**
     * Item template
     * @type {?}
     */
    TypeaheadOptionListContext.prototype.itemTemplate;
    /**
     * Search query
     * @type {?}
     */
    TypeaheadOptionListContext.prototype.query;
    /**
     * Typeahead template methods
     * @type {?}
     */
    TypeaheadOptionListContext.prototype.$implicit;
}
/**
 * A context for the `typeaheadItemTemplate`
 * input template in case you want to override default one
 * @record
 */
export function TypeaheadOptionItemContext() { }
if (false) {
    /**
     * Item
     * @type {?}
     */
    TypeaheadOptionItemContext.prototype.item;
    /**
     * Item index
     * @type {?}
     */
    TypeaheadOptionItemContext.prototype.index;
    /**
     * Typeahead match
     * @type {?}
     */
    TypeaheadOptionItemContext.prototype.match;
    /**
     * Search query
     * @type {?}
     */
    TypeaheadOptionItemContext.prototype.query;
}
/**
 * Methods for `optionsListTemplate` context
 * @record
 */
export function TypeaheadTemplateMethods() { }
if (false) {
    /**
     * Function to select an option by click event
     * @param {?} value
     * @param {?=} e
     * @return {?}
     */
    TypeaheadTemplateMethods.prototype.selectMatch = function (value, e) { };
    /**
     * Function to select an option by mouseenter event
     * @param {?} value
     * @return {?}
     */
    TypeaheadTemplateMethods.prototype.selectActive = function (value) { };
    /**
     * Function to check if an option is active
     * @param {?} value
     * @return {?}
     */
    TypeaheadTemplateMethods.prototype.isActive = function (value) { };
}
