// @angular/localize/src/tools/src/extract/translation_files/icu_parsing.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/localize/src/tools/src/extract/translation_files/icu_parsing", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.extractIcuPlaceholders = void 0;
    /**
     * Split the given `text` into an array of "static strings" and ICU "placeholder names".
     *
     * This is required because ICU expressions in `$localize` tagged messages may contain "dynamic"
     * piece (e.g. interpolations or element markers). These markers need to be translated to
     * placeholders in extracted translation files. So we must parse ICUs to identify them and separate
     * them out so that the translation serializers can render them appropriately.
     *
     * An example of an ICU with interpolations:
     *
     * ```
     * {VAR_PLURAL, plural, one {{INTERPOLATION}} other {{INTERPOLATION_1} post}}
     * ```
     *
     * In this ICU, `INTERPOLATION` and `INTERPOLATION_1` are actually placeholders that will be
     * replaced with dynamic content at runtime.
     *
     * Such placeholders are identifiable as text wrapped in curly braces, within an ICU case
     * expression.
     *
     * To complicate matters, it is possible for ICUs to be nested indefinitely within each other. In
     * such cases, the nested ICU expression appears enclosed in a set of curly braces in the same way
     * as a placeholder. The nested ICU expressions can be differentiated from placeholders as they
     * contain a comma `,`, which separates the ICU value from the ICU type.
     *
     * Furthermore, nested ICUs can have placeholders of their own, which need to be extracted.
     *
     * An example of a nested ICU containing its own placeholders:
     *
     * ```
     * {VAR_SELECT_1, select,
     *   invoice {Invoice for {INTERPOLATION}}
     *   payment {{VAR_SELECT, select,
     *     processor {Payment gateway}
     *     other {{INTERPOLATION_1}}
     *   }}
     * ```
     *
     * @param text Text to be broken.
     * @returns an array of strings, where
     *  - even values are static strings (e.g. 0, 2, 4, etc)
     *  - odd values are placeholder names (e.g. 1, 3, 5, etc)
     */
    function extractIcuPlaceholders(text) {
        var state = new StateStack();
        var pieces = new IcuPieces();
        var braces = /[{}]/g;
        var lastPos = 0;
        var match;
        while (match = braces.exec(text)) {
            if (match[0] == '{') {
                state.enterBlock();
            }
            else {
                // We must have hit a `}`
                state.leaveBlock();
            }
            if (state.getCurrent() === 'placeholder') {
                var name = tryParsePlaceholder(text, braces.lastIndex);
                if (name) {
                    // We found a placeholder so store it in the pieces;
                    // store the current static text (minus the opening curly brace);
                    // skip the closing brace and leave the placeholder block.
                    pieces.addText(text.substring(lastPos, braces.lastIndex - 1));
                    pieces.addPlaceholder(name);
                    braces.lastIndex += name.length + 1;
                    state.leaveBlock();
                }
                else {
                    // This is not a placeholder, so it must be a nested ICU;
                    // store the current static text (including the opening curly brace).
                    pieces.addText(text.substring(lastPos, braces.lastIndex));
                    state.nestedIcu();
                }
            }
            else {
                pieces.addText(text.substring(lastPos, braces.lastIndex));
            }
            lastPos = braces.lastIndex;
        }
        // Capture the last piece of text after the ICUs (if any).
        pieces.addText(text.substring(lastPos));
        return pieces.toArray();
    }
    exports.extractIcuPlaceholders = extractIcuPlaceholders;
    /**
     * A helper class to store the pieces ("static text" or "placeholder name") in an ICU.
     */
    var IcuPieces = /** @class */ (function () {
        function IcuPieces() {
            this.pieces = [''];
        }
        /**
         * Add the given `text` to the current "static text" piece.
         *
         * Sequential calls to `addText()` will append to the current text piece.
         */
        IcuPieces.prototype.addText = function (text) {
            this.pieces[this.pieces.length - 1] += text;
        };
        /**
         * Add the given placeholder `name` to the stored pieces.
         */
        IcuPieces.prototype.addPlaceholder = function (name) {
            this.pieces.push(name);
            this.pieces.push('');
        };
        /**
         * Return the stored pieces as an array of strings.
         *
         * Even values are static strings (e.g. 0, 2, 4, etc)
         * Odd values are placeholder names (e.g. 1, 3, 5, etc)
         */
        IcuPieces.prototype.toArray = function () {
            return this.pieces;
        };
        return IcuPieces;
    }());
    /**
     * A helper class to track the current state of parsing the strings for ICU placeholders.
     *
     * State changes happen when we enter or leave a curly brace block.
     * Since ICUs can be nested the state is stored as a stack.
     */
    var StateStack = /** @class */ (function () {
        function StateStack() {
            this.stack = [];
        }
        /**
         * Update the state upon entering a block.
         *
         * The new state is computed from the current state and added to the stack.
         */
        StateStack.prototype.enterBlock = function () {
            var current = this.getCurrent();
            switch (current) {
                case 'icu':
                    this.stack.push('case');
                    break;
                case 'case':
                    this.stack.push('placeholder');
                    break;
                case 'placeholder':
                    this.stack.push('case');
                    break;
                default:
                    this.stack.push('icu');
                    break;
            }
        };
        /**
         * Update the state upon leaving a block.
         *
         * The previous state is popped off the stack.
         */
        StateStack.prototype.leaveBlock = function () {
            return this.stack.pop();
        };
        /**
         * Update the state upon arriving at a nested ICU.
         *
         * In this case, the current state of "placeholder" is incorrect, so this is popped off and the
         * correct "icu" state is stored.
         */
        StateStack.prototype.nestedIcu = function () {
            var current = this.stack.pop();
            assert(current === 'placeholder', 'A nested ICU must replace a placeholder but got ' + current);
            this.stack.push('icu');
        };
        /**
         * Get the current (most recent) state from the stack.
         */
        StateStack.prototype.getCurrent = function () {
            return this.stack[this.stack.length - 1];
        };
        return StateStack;
    }());
    /**
     * Attempt to parse a simple placeholder name from a curly braced block.
     *
     * If the block contains a comma `,` then it cannot be a placeholder - and is probably a nest ICU
     * instead.
     *
     * @param text the whole string that is being parsed.
     * @param start the index of the character in the `text` string where this placeholder may start.
     * @returns the placeholder name or `null` if it is not a placeholder.
     */
    function tryParsePlaceholder(text, start) {
        for (var i = start; i < text.length; i++) {
            if (text[i] === ',') {
                break;
            }
            if (text[i] === '}') {
                return text.substring(start, i);
            }
        }
        return null;
    }
    function assert(test, message) {
        if (!test) {
            throw new Error('Assertion failure: ' + message);
        }
    }
});
