// ngx-bootstrap/chronos/esm5/parse/regex.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { hasOwnProp, isFunction } from '../utils/type-checks';
/** @type {?} */
export var match1 = /\d/;
//       0 - 9
/** @type {?} */
export var match2 = /\d\d/;
//      00 - 99
/** @type {?} */
export var match3 = /\d{3}/;
//     000 - 999
/** @type {?} */
export var match4 = /\d{4}/;
//    0000 - 9999
/** @type {?} */
export var match6 = /[+-]?\d{6}/;
// -999999 - 999999
/** @type {?} */
export var match1to2 = /\d\d?/;
//       0 - 99
/** @type {?} */
export var match3to4 = /\d\d\d\d?/;
//     999 - 9999
/** @type {?} */
export var match5to6 = /\d\d\d\d\d\d?/;
//   99999 - 999999
/** @type {?} */
export var match1to3 = /\d{1,3}/;
//       0 - 999
/** @type {?} */
export var match1to4 = /\d{1,4}/;
//       0 - 9999
/** @type {?} */
export var match1to6 = /[+-]?\d{1,6}/;
// -999999 - 999999
/** @type {?} */
export var matchUnsigned = /\d+/;
//       0 - inf
/** @type {?} */
export var matchSigned = /[+-]?\d+/;
//    -inf - inf
/** @type {?} */
export var matchOffset = /Z|[+-]\d\d:?\d\d/gi;
// +00:00 -00:00 +0000 -0000 or Z
/** @type {?} */
export var matchShortOffset = /Z|[+-]\d\d(?::?\d\d)?/gi;
// +00 -00 +00:00 -00:00 +0000 -0000 or Z
/** @type {?} */
export var matchTimestamp = /[+-]?\d+(\.\d{1,3})?/;
// 123456789 123456789.123
// any word (or two) characters or numbers including two/three word month in arabic.
// includes scottish gaelic two word and hyphenated months
// tslint:disable-next-line
/** @type {?} */
export var matchWord = /[0-9]{0,256}['a-z\u00A0-\u05FF\u0700-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]{1,256}|[\u0600-\u06FF\/]{1,256}(\s*?[\u0600-\u06FF]{1,256}){1,2}/i;
/** @type {?} */
var regexes = {};
/**
 * @param {?} token
 * @param {?} regex
 * @param {?=} strictRegex
 * @return {?}
 */
export function addRegexToken(token, regex, strictRegex) {
    if (isFunction(regex)) {
        regexes[token] = regex;
        return;
    }
    regexes[token] = (/**
     * @param {?} isStrict
     * @param {?} locale
     * @return {?}
     */
    function (isStrict, locale) {
        return (isStrict && strictRegex) ? strictRegex : regex;
    });
}
/**
 * @param {?} token
 * @param {?} locale
 * @return {?}
 */
export function getParseRegexForToken(token, locale) {
    /** @type {?} */
    var _strict = false;
    if (!hasOwnProp(regexes, token)) {
        return new RegExp(unescapeFormat(token));
    }
    return regexes[token](_strict, locale);
}
// Code from http://stackoverflow.com/questions/3561493/is-there-a-regexp-escape-function-in-javascript
/**
 * @param {?} str
 * @return {?}
 */
function unescapeFormat(str) {
    // tslint:disable-next-line
    return regexEscape(str
        .replace('\\', '')
        .replace(/\\(\[)|\\(\])|\[([^\]\[]*)\]|\\(.)/g, (/**
     * @param {?} matched
     * @param {?} p1
     * @param {?} p2
     * @param {?} p3
     * @param {?} p4
     * @return {?}
     */
    function (matched, p1, p2, p3, p4) { return p1 || p2 || p3 || p4; })));
}
/**
 * @param {?} str
 * @return {?}
 */
export function regexEscape(str) {
    return str.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
}
